### PR TITLE
feat: Check for allocation expiry when waiting to seal sectors

### PR DESCRIPTION
**Pull request by geoff-vball**
## Related Issues
Closes https://github.com/filecoin-project/lotus/issues/9368

## Proposed Changes
Checks for allocation expirations to make sure we don't wait too long before sealing. This is slightly tricky because allocations are allowed to expire, the miner will just not get the increased power for the verified deal.

## Additional Info
This could use some testing. 

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by geoff-vball** [comment]
Currently failing sector_prefer_no_upgrade
 https://app.circleci.com/pipelines/github/filecoin-project/lotus/25014/workflows/259d1b44-c6b8-46b1-a5ff-2d584e156091/jobs/719886/tests#failed-test-0

File: storage/pipeline/input.go [hunk 1]
```diff
@@ -143,17 +143,34 @@ func (m *Sealing) maybeStartSealing(ctx statemachine.Context, sector SectorInfo,
...
+			if err != nil {
+				return false, err
+			}
+			// alloc is nil if this is not a verified deal
+			if alloc == nil {
+				continue
+			}
+
+			if alloc.Expiration-cfg.StartEpochSealingBuffer < dealSafeSealEpoch {
+				dealSafeSealEpoch = alloc.Expiration - cfg.StartEpochSealingBuffer
```
Comment (magik6k) [inline storage/pipeline/input.go:165]: Could add a `log.Debugw` here to have some way to debug/see if it works this when deployed in the wild

**Comment by magik6k** [comment]
Will need rebase after https://github.com/filecoin-project/lotus/pull/9941 is merged

**Comment by geoff-vball** [comment]
@magik6k I've fixed the test that was failing here and it's ready for another look.

**Review by magik6k — APPROVED** [review]

**Pull request merged**